Exporting functions with parameter initialisers in interfaces fails TS compilation

If a class is exported as interface and class' methods have parameter initialisers, these initialisers need to be omitted. TypeScript fails compilation for methods in interfaces that have parameter initialisers.

So, parameter initialisers need to be omitted from methods in case the type is exported as an interface.

## TypeScript compiler error

The following error is created by TypeScript compiler (version 3.9.7)
```
src/assets.ts:26:31 - error TS2371: A parameter initializer is only allowed in a function or constructor implementation.

26     getRandomImageAssetsAsync(max: number = 10) : Promise<IAssetDataObject[]>;
                                 ~~~~~~~~~~~~~~~~
```

## source files

The file has been exported from C# and the failing TypeScript interface looks like:
```typesscript
export interface IAssetsReadRandom
{
    getRandomImageAssetsAsync(max: number = 10) : Promise<IAssetDataObject[]>;
}
```

This is the interface from C#
```C#
    public interface IAssetsReadRandom
    {
        Task<AssetDataObject[]> GetRandomImageAssetsAsync(int max = 10);
    }
````

## Test case - expected result

A default value (initial value) makes the parameter optional.
When exporting such a method to an interface, it must be omitted from the interface definition. Nevertheless, the parameter still is optional as the default value is in place with the implementation. Hence, the parameter should be marked as "nullable" with "?".

```typesscript
export interface IAssetsReadRandom
{
    getRandomImageAssetsAsync(max?: number) : Promise<IAssetDataObject[]>;
}
```


The following test case will highlight the error:

```C#
using System.Reflection;
using Reinforced.Typings.Ast;
using Reinforced.Typings.Attributes;
using Reinforced.Typings.Fluent;
using Reinforced.Typings.Generators;
using Xunit;

namespace Reinforced.Typings.Tests.SpecificCases
{
    public partial class SpecificTestCases
    {
        private class MyFunctionTestClassWithParamInitializer
        {
        
            public string MyProperty { get; set; }
        
            public int MyNumber { get; set; }

            public string DoSomething(int a = 4)
            {
                return a == 4 ? "Hello World" : string.Empty;
            }
        }

        [Fact]
        public void FunctionWithParamInitializerAsInterface()
        {
            const string result = @"
                module Reinforced.Typings.Tests.SpecificCases {
	                export interface IMyFunctionTestClassWithParamInitializer
	                {
		                MyProperty: string;
		                MyNumber: number;
		                doSomething(a?: number) : string;
	                }
                }
            ";

            AssertConfiguration(s =>
            {
                s.Global(x=>x.DontWriteWarningComment());
                s.ExportAsInterface<MyFunctionTestClassWithParamInitializer>()
                .WithPublicMethods(x=>x.CamelCase())
                .WithPublicProperties();
            }, result);
        }
    }
}
```


## possible cure

The following change would emit parameter initialisers only if the write context is not an interface.
```diff
--- a/Reinforced.Typings/Visitors/TypeScript/TypeScriptExportVisitor.RtArgument.cs
+++ b/Reinforced.Typings/Visitors/TypeScript/TypeScriptExportVisitor.RtArgument.cs
@@ -10,9 +10,16 @@ namespace Reinforced.Typings.Visitors.TypeScript
             Decorators(node);
             if (node.IsVariableParameters) Write("...");
             Visit(node.Identifier);
+
+            // if a default value is used, then the parameter may be omitted as the default value is in place.
+            if (Context == WriterContext.Interface && !node.Identifier.IsNullable && !string.IsNullOrEmpty(node.DefaultValue))
+            {
+                Write("?");
+            }
+
             Write(": ");
             Visit(node.Type);
-            if (!string.IsNullOrEmpty(node.DefaultValue))
+            if (Context != WriterContext.Interface && !string.IsNullOrEmpty(node.DefaultValue))
             {
                 Write(" = ");
                 Write(node.DefaultValue);
```

--- a/Reinforced.Typings.Tests/SpecificCases/SpecificTestCases.FunctionWithParamInitialiser.cs
+++ b/Reinforced.Typings.Tests/SpecificCases/SpecificTestCases.FunctionWithParamInitialiser.cs
@@ -1,0 +1,73 @@
+﻿using System.Reflection;
+using Reinforced.Typings.Ast;
+using Reinforced.Typings.Attributes;
+using Reinforced.Typings.Fluent;
+using Reinforced.Typings.Generators;
+using Xunit;
+
+namespace Reinforced.Typings.Tests.SpecificCases
+{
+    public partial class SpecificTestCases
+    {
+        private class MyFunctionTestClassWithParamInitializer
+        {
+        
+            public string MyProperty { get; set; }
+        
+            public int MyNumber { get; set; }
+
+            public string DoSomething(int a = 4)
+            {
+                return a == 4 ? "Hello World" : string.Empty;
+            }
+        }
+
+        [Fact]
+        public void FunctionWithParamInitializerAsInterface()
+        {
+            const string result = @"
+                module Reinforced.Typings.Tests.SpecificCases {
+	                export interface IMyFunctionTestClassWithParamInitializer
+	                {
+		                MyProperty: string;
+		                MyNumber: number;
+		                doSomething(a?: number) : string;
+	                }
+                }
+            ";
+
+            AssertConfiguration(s =>
+            {
+                s.Global(x=>x.DontWriteWarningComment());
+                s.ExportAsInterface<MyFunctionTestClassWithParamInitializer>()
+                .WithPublicMethods(x=>x.CamelCase())
+                .WithPublicProperties();
+            }, result);
+        }
+
+        [Fact]
+        public void FunctionWithParamInitializerAsClass()
+        {
+            const string result = @"
+                module Reinforced.Typings.Tests.SpecificCases {
+	                export class MyFunctionTestClassWithParamInitializer
+	                {
+		                public MyProperty: string;
+		                public MyNumber: number;
+		                public doSomething(a: number = 4) : string
+		                {
+			                return null;
+		                }
+	                }
+                }
+                ";
+            AssertConfiguration(s =>
+            {
+                s.Global(x=>x.DontWriteWarningComment());
+                s.ExportAsClass<MyFunctionTestClassWithParamInitializer>()
+                    .WithPublicMethods(x=>x.CamelCase())
+                    .WithPublicProperties();
+            }, result);
+        }
+    }
+}

--- a/Reinforced.Typings/Visitors/TypeScript/TypeScriptExportVisitor.RtArgument.cs
+++ b/Reinforced.Typings/Visitors/TypeScript/TypeScriptExportVisitor.RtArgument.cs
@@ -10,9 +10,16 @@ namespace Reinforced.Typings.Visitors.TypeScript
             Decorators(node);
             if (node.IsVariableParameters) Write("...");
             Visit(node.Identifier);
+
+            // if a default value is used, then the parameter may be omitted as the default value is in place.
+            if (Context == WriterContext.Interface && !node.Identifier.IsNullable && !string.IsNullOrEmpty(node.DefaultValue))
+            {
+                Write("?");
+            }
+
             Write(": ");
             Visit(node.Type);
-            if (!string.IsNullOrEmpty(node.DefaultValue))
+            if (Context != WriterContext.Interface && !string.IsNullOrEmpty(node.DefaultValue))
             {
                 Write(" = ");
                 Write(node.DefaultValue);

